src/Microsoft.Azure.SignalR/EndpointRouters/DefaultEndpointRouter.cs

```diff
@@ -45,5 +46,38 @@ private ServiceEndpoint[] GetNegotiateEndpoints(IEnumerable<ServiceEndpoint> end
 
             return secondary;
         }
+
+        private ServiceEndpoint GetEndpointAccordingToWeight(ServiceEndpoint[] availableEndpoints)
+        {
+            //first check if weight is available or necessary
+            if (availableEndpoints.Any(endpoint => endpoint.EndpointMetrics.ConnectionCapacity == 0) ||
+                availableEndpoints.Length == 1)
+            {
+                return GetEndpointRandomly(availableEndpoints);
+            }
+
+            var we = new int[availableEndpoints.Length];
+            var totalCapacity = 0;
+            for (var i = 0; i < availableEndpoints.Length; i++)
+            {
+                var endpointMetrics = availableEndpoints[i].EndpointMetrics;
+                var remain = endpointMetrics.ConnectionCapacity -
+                             (endpointMetrics.ClientConnectionCount +
+                              endpointMetrics.ServerConnectionCount);
+                var weight = remain > 0 ? remain : 1;
+                totalCapacity += weight;
+                we[i] = totalCapacity;
+            }
+
+            var index = StaticRandom.Next(totalCapacity);
+            var key = Array.BinarySearch(we, index);
```

Comment:
there is no need to use `BinarySearch`, in normal case the `availableEndpoints` should be than 10.

Reply:
Change to a for loop